server/service/src/invoice/inventory_adjustment/insert.rs

```diff
@@ -0,0 +1,245 @@
+        .map_err(|error| error.to_inner_error())?;
+
+    Ok(invoice)
+}
+
+impl From<RepositoryError> for InsertInventoryAdjustmentError {
+    fn from(error: RepositoryError) -> Self {
+        InsertInventoryAdjustmentError::DatabaseError(error)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use repository::{
+        mock::{
+            mock_stock_line_a, mock_store_a, mock_store_b, mock_user_account_a, MockDataInserts,
+        },
+        test_db::setup_all,
+        InvoiceRowRepository, InvoiceRowStatus, StockLine,
+    };
+    use util::inline_edit;
+
+    use crate::{
+        invoice::inventory_adjustment::insert::InsertInventoryAdjustment,
+        service_provider::ServiceProvider,
+    };
+
+    use super::InsertInventoryAdjustmentError;
+
+    type ServiceError = InsertInventoryAdjustmentError;
+
+    #[actix_rt::test]
+    async fn insert_inventory_adjustment_errors() {
+        let (_, _, connection_manager, _) =
+            setup_all("insert_inventory_adjustment_errors", MockDataInserts::all()).await;
+
+        let service_provider = ServiceProvider::new(connection_manager, "app_data");
+        let mut context = service_provider
+            .context(mock_store_a().id, "".to_string())
+            .unwrap();
+        let service = service_provider.invoice_service;
+
+        // Stockline doesn't exist
+        assert_eq!(
+            service.insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: "x".to_string(),
+                    ..Default::default()
+                }
+            ),
+            Err(ServiceError::StockLineDoesNotExist)
+        );
+
+        // Wrong store
+        context.store_id = mock_store_b().id;
+        assert_eq!(
+            service.insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: mock_stock_line_a().id,
+                    ..Default::default()
+                }
+            ),
+            Err(ServiceError::InvalidStore)
+        );
+        context.store_id = mock_store_a().id;
+
+        // Missing reason
+        // assert_eq!(
+        //     service.insert_inventory_adjustment(
+        //         &context,
+        //         InsertInventoryAdjustment {
+        //             stock_line_id: mock_stock_line_a().id,
+        //             ..Default::default()
+        //         }
+        //     ),
+        //     Err(ServiceError::AdjustmentReasonNotProvided)
+        // );
+
+        // Invalid reason
+        assert_eq!(
+            service.insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: mock_stock_line_a().id,
+                    inventory_adjustment_reason_id: Some("invalid".to_string()),
+                    ..Default::default()
+                }
+            ),
+            Err(ServiceError::InvalidAdjustment)
```

Comment:
Should be `AdjustmentReasonNotValid`... would need to provide adjustment & adjustment_type for it to get past the `input.adjustment == 0.0` check though!